approximate_string_matching/four_russians_helpers.py

```diff
@@ -0,0 +1,260 @@
+
+    return storage
+
+  def compute_next_step_vectors(self, C, D, T, U, m):
+    for i in range(1, m + 1):
+      for j in range(1, m + 1):
+        T[i][j] = min(
+            self.substitute_cost_function(C[i], D[j]) - U[i - 1][j],
+            self.delete_cost_function(C[i]),
+            self.insert_cost_function(D[j]) + T[i][j - 1] - U[i - 1][j])
+        U[i][j] = min(
+            self.substitute_cost_function(C[i], D[j]) - T[i][j - 1],
+            self.delete_cost_function(C[i]) + U[i - 1][j] - T[i][j - 1],
+            self.insert_cost_function(D[j])
+        )
+
+  def algorithm_y(self, m, A, step_size_bound):
+    """ Preprocesses data by creating helper submatrices """
+
+    if len(A) == 0:
+      return []
+
+    A = ['#'] + A
+    strings = self.get_all_strings(m, A, '#')
+
+    step_vectors_alphabet = [[cost] for cost in
+                             range(-step_size_bound, step_size_bound + 1)]
+    step_vectors = self.get_all_strings(m, step_vectors_alphabet, [0])
+
+    return self.generate_final_step_vectors(m, step_vectors, strings)
+
+  @staticmethod
+  def get_text_parts(m, text_1):
+    return int((len(text_1) - 1) / m)
+
+  def algorithm_z(self, m, storage, text_1, text_2):
+    """ Returns step vectors needed to calculate the edit distance
+    between A and B using preprocessed submatrices of size mxm """
+
+    text_1_parts_bound = self.get_text_parts(m, text_1) + 1
+    text_2_parts_bound = self.get_text_parts(m, text_2) + 1
+
+    P = [[[] for _ in range(text_2_parts_bound)] for _ in
+         range(text_1_parts_bound)]
+
+    for i in range(1, text_1_parts_bound):
+      P[i][0] = [self.delete_cost_function(text_1[letter_idx]) for letter_idx in
+                 range((i - 1) * m + 1, i * m + 1)]
+
+    Q = [[[] for _ in range(text_2_parts_bound)] for _ in
+         range(text_1_parts_bound)]
+
+    for j in range(1, text_2_parts_bound):
+      Q[0][j] = [self.insert_cost_function(text_2[letter_idx]) for letter_idx in
+                 range((j - 1) * m + 1, j * m + 1)]
+
+    for i in range(1, text_1_parts_bound):
+      for j in range(1, text_2_parts_bound):
+        (P[i][j], Q[i][j]) = self.fetch(
+            self.get_kth_substring(i, m, text_1),
+            self.get_kth_substring(j, m, text_2),
+            P[i][j - 1], Q[i - 1][j],
+            storage)
+
+    return P, Q
+
+  def restore_matrix(self, C, D, R, S, m):
+    M = [[0] * (m + 1) for _ in range(m + 1)]
+    for i in range(1, m + 1):
+      M[i][0] = R[i]
+      M[0][i] = S[i]
+    for i in range(1, m + 1):
+      for j in range(1, m + 1):
+        M[i][j] = min(
+            self.substitute_cost_function(C[i], D[j]) + M[i - 1][j - 1],
+            self.delete_cost_function(C[i]) + M[i - 1][j],
+            self.insert_cost_function(D[j]) + M[i][j - 1]
+        )
+    return M
+
+  def restore_lcs_part(self, C, D, R, S, m, i1, j1):
+    M = self.restore_matrix(C, D, R, S, m)
+
+    i = i1
+    j = j1
```

Comment:
A natural place for a multiple assignment. Also it may be applied in many other places.